src/hypertable.c

```diff
@@ -1918,7 +1922,33 @@ ts_hypertable_create_internal(PG_FUNCTION_ARGS, bool is_dist_call)
 	ht = ts_hypertable_cache_get_cache_and_entry(table_relid, CACHE_FLAG_NONE, &hcache);
 
 	if (NULL != space_dim_info)
+	{
```

Comment:
I don't quite understand the change. If the table is not created , why do we validate the space_dim_info here (i.e. the check at 1950)? Shoudln't it skip this check all together? space_dim_info was validated when the table was created.

Reply:
In theory, the check should only be necessary when the table is created, but just running the check when the table is created still causes a crash. Conditioning the check on the `space_dim_info != NULL` runs the risk of not doing the check when it should be done.

The code is in need of a refactoring but I would like to avoid that since there is a risk of a snowball effect. There are several issues that we should refactor away:

* Push down the partitioning check into `ts_hypertable_create_from_info`.
* Ensure that validation does not change the space dimension structure. This will require moving the setting of, among other things, the `dimension_id` of the structure to somewhere else, potentially altering the logic of the function.
* Remove the validation if the table is not created. Right now, there are checks done at at least two places, but it requires a little more research to figure out the good way.
* Dimensions are partially created, as in this case, without filling in all the information from the dimension table. Adding redundant loads are not a good idea, since it affect performance, but making sure that we load all the necessary information when and only when needed requires some more research.

Reply:
I also find this change odd and I worry that this is adding more confusion. Isn't it better to add an early check for existing hypertable at the beginning of the function and then skipping all the creation stuff in that case, thus skipping to the end?

In other words, add a call to `ts_hypertable_cache_get_cache_and_entry` at the start of the function and then immediately skip to the end (`create_hypertable_datum(fcinfo, ht, created)` ) if the hypertable exist.

Reply:
If I add an early check for the hypertable and skip all the creation code, I get the following result when running regression tests:

```
test bgw_db_scheduler             ... ok
test bgw_launcher                 ... ok
test alternate_users              ... FAILED
test net                          ... ok
test pg_dump_unprivileged         ... ok
test telemetry                    ... FAILED
parallel group (20 tests):  chunk_utils_compression copy c_unit_tests chunks cursor chunk_utils create_hypertable constraint ddl_errors create_chunks agg_bookends-11 chunk_adaptive-11 create_table append-11 ddl_alter_column ddl_single-11 custom_type-11 alter cluster ddl-11
     agg_bookends-11              ... FAILED
     alter                        ... FAILED
     append-11                    ... FAILED
     c_unit_tests                 ... ok
     chunk_adaptive-11            ... FAILED
     chunk_utils                  ... FAILED
     chunk_utils_compression      ... FAILED
     chunks                       ... ok
     cluster                      ... FAILED
     constraint                   ... FAILED
     copy                         ... FAILED
     create_chunks                ... FAILED
     create_hypertable            ... FAILED
     create_table                 ... FAILED
     cursor                       ... FAILED
     custom_type-11               ... FAILED
     ddl-11                       ... FAILED
     ddl_alter_column             ... FAILED
     ddl_errors                   ... FAILED
     ddl_single-11                ... FAILED
parallel group (20 tests):  drop_extension drop_hypertable delete insert_many drop_rename_hypertable lateral drop_schema gapfill generated_as_identity insert-11 guc_options drop_owned hash extension index histogram_test join insert_single dump_meta edition
     delete                       ... FAILED
     drop_extension               ... FAILED
     drop_hypertable              ... FAILED
     drop_owned                   ... FAILED
     drop_rename_hypertable       ... FAILED
     drop_schema                  ... FAILED
     dump_meta                    ... FAILED
     edition                      ... ok
     extension                    ... ok
     gapfill                      ... ok
     generated_as_identity        ... FAILED
     guc_options                  ... ok
     hash                         ... ok
     histogram_test               ... ok
     index                        ... FAILED
     insert-11                    ... FAILED
     insert_many                  ... FAILED
     insert_single                ... FAILED
     join                         ... FAILED
     lateral                      ... FAILED
parallel group (20 tests):  multi_transaction_index-11 plan_hashagg-11 partition-11 plan_ordered_append parallel-11 partitioning reloptions partitionwise size_utils plain plan_expand_hypertable-11 pg_dump query-11 reindex plan_hypertable_cache-11 privacy relocate_extension rowsecurity-11 loader metadata
     loader                       ... ok
     metadata                     ... ok
     multi_transaction_index-11   ... FAILED
     parallel-11                  ... FAILED
     partition-11                 ... FAILED
     partitioning                 ... FAILED
     partitionwise                ... FAILED
     pg_dump                      ... FAILED
     plain                        ... ok
     plan_expand_hypertable-11    ... FAILED
     plan_hashagg-11              ... FAILED
     plan_hypertable_cache-11     ... FAILED
     plan_ordered_append          ... FAILED
     privacy                      ... ok
     query-11                     ... FAILED
     reindex                      ... FAILED
     relocate_extension           ... FAILED
     reloptions                   ... FAILED
     rowsecurity-11               ... FAILED
     size_utils                   ... FAILED
parallel group (17 tests):  version sql_query-11 vacuum vacuum_multi timestamp sort_optimization-11 upsert truncate tablespace triggers util views update symbol_conflict telemetry_community test_utils telemetry_compression
     sort_optimization-11         ... FAILED
     sql_query-11                 ... FAILED
     symbol_conflict              ... ok
     tablespace                   ... FAILED
     telemetry_community          ... FAILED
     telemetry_compression        ... FAILED
     test_utils                   ... ok
     timestamp                    ... FAILED
     triggers                     ... FAILED
     truncate                     ... FAILED
     update                       ... FAILED
     upsert                       ... FAILED
     util                         ... ok
     vacuum                       ... FAILED
     vacuum_multi                 ... FAILED
     version                      ... ok
     views                        ... FAILED
============== shutting down postmaster               ==============

========================
 63 of 83 tests failed. 
========================
```

If I only check if the table was created after the call, I still get the crash. I haven't investigated further. I think trying to do it "properly" at this time is a significant amount of work and require quite a lot of refactoring, which I do not think is warranted at this time.

Reply:
I don't understand why an early check for the existence of a hypertable wouldn't work---it doesn't make any sense that it wouldn't. I changed a few lines of code (note that most lines changed are indentation): https://github.com/erimatnor/timescaledb/commit/2c9d04b5a57feb9586f327cb2f274cda6d9c2021

and it passes all tests. Not sure why it failed for you.

Reply:
I added your suggested fix to get a test run for the PR and see if I get the same result as I do locally. You suggested fix is close to what I used for the test run above, but not identical.

The test results in the checks below are the same in the checks as on my machine locally.

I am not sure how you got it to pass.